### ISSUE TITLE
Handling of external libraries

### ISSUE THREAD

**Issue by cochrane** [opentomb/OpenTomb#115]
We need to converge with this discussion somewhere because right now, it is all over the place. So here are the possibilities for handling the dependencies our code has that I've been able to think of:

-   Store the code directly in the repository. Pro: Very easy; no setup required; inherently works on all platforms. Can make own modifications. Cons: Can make own modifications that aren't visible to the outside world; update difficult.
-   Store the binaries directly. Pro: Update becomes easy. Con: Need at least three versions of everything (Mac, Linux, Windows), possibly more depending on 32/64 bit. Any new platform (does FreeBSD have OpenGL?) increases the pain.
-   Hope the user has installed the libraries. Pro: Nothing to do on our part except edit the Readme.md. Con: No idea who uses what version of what library.
-   Use external Git references to the libraries (to known tags). Pro: Using official version; no modifications make it back. Same ease of use as storing in repository. Con: Git external references are really hard. Especially if some library we want doesn't have a git repository; there's some Git/SVN bridge but I've never tried it.

A few more notes:
-   SDL is funny because the Mac version is a different format (Framework) than any other library. That's my duty to figure out, of course, for everyone else that's not an issue.
-   How big a problem is incompatibility really? Probably not that big; incompatible OpenGL drivers hurt way more. But of course it takes only one issue to make the whole thing less fun.
-   We definitely need some way to deal with our own version of OpenAL. It does things the normal version doesn't, and I don't want to throw that away without very good reason. We could put it in a separate repository here and then use it like any other third-party library, of course.
-   My personal vote would be for Git external references, I just have very little experience with actually using them.

**Comment by Lwmte** [comment]
Cochrane, is it possible to install/compile/link OpenAL Soft (http://kcat.strangesoft.net/openal.html) for Mac along with stock OpenAL? Bundled version isn't very different from default OpenAL Soft, except it used custom-written SDL backend (which I will really miss - but I wrote a letter to Chris Robinson asking if he can implement SDL backend officially, or at least XAudio2 output for Windows).

I have more concerns regarding recently introduced **sndfile** library. I won't babble about how easy and simple it was without it and with only using native SDL functions to read and decompress audio samples, but what is the whole point in using sndfile *along with* OGG/Vorbis, if sndfile supports OGG/Vorbis itself?! Basically, we now have FOUR sound library dependencies (SDL, OpenAL Soft, OGG/Vorbis and now sndfile)! I'm sorry, but OpenTomb isn't a sound editing or video editing software, we need it light and simple - so either get rid of OGG/Vorbis or sndfile.

By the way, if sndfile can provide EASY interface for streaming MS-ADPCM wave files (TR3-TR5), then it should absolutely stay in place. I just had no idea how to program MS-ADPCM streaming when I wrote audio engine.

**Comment by Lwmte** [comment]
P.S.: Sorry, but maybe I got something wrong... Maybe sndfile **requires** OGG/Vorbis lib to read .ogg files? If that's the case, then self-written OGG streaming code should go away, and sndfile streaming should be implemented instead.

**Comment by stohrendorf** [comment]
MS-ADPCM is possible (see http://www.mega-nerd.com/libsndfile/); and from what I can tell from the sources (https://github.com/erikd/libsndfile/tree/master/src), it requires *external* libogg and libvorbis, so we still need to link against it.

**Comment by Lwmte** [comment]
Ok then, I thought libsndfile supports OGG/Vorbis itself! Maybe using libsndfile for streaming will solve current loading lag issue with soundtracks.

Generally - I'm all for removing third-party libs from source code, and using third approach offered by Cochrane - just put an info about version requirements into Readme, and hope that user will install libraries for himself. Because it's already happening anyway (I needed to install SDL and zlib anyway, and in case of zlib, I had funny time finding pre-compiled libs on the internet, as I'm too lazy and dumb to compile them myself).

Not sure if this approach will break nightly builds, we need to ask vobject about that... So far it worked with external SDL and zlib.

**Comment by stohrendorf** [comment]
Maybe we should set up a wiki page with dev requirements, like limiting boost to only use headers, external library handling and versioning, coding style, whatever.

**Comment by cochrane** [comment]
The wiki page is a good idea to document whatever we come up with here, once we have something. 

@Lwmte: Installing a custom version of OpenAL is no problem on OS X, and I think actually what most games do. But again, throwing out our existing patched version needs strong justification that I just don't see.

**Comment by cochrane** [comment]
I'm not fond of having everyone install their own libraries. Everything that makes it more difficult for new developers to join is a problem, not an advantage.

**Comment by vvs-** [comment]
Actually, I see two problems here:

* a development environment
* a deployment environment

They can have different setups. Actually, there is a special MinGW case already for cross-compilation.

**Comment by stohrendorf** [comment]
To summarize: my last merge request eliminated the need for SDL Audio, so the only depencies are now OpenAL and libsndfile (which in turn requires ogg/vorbis). Both libraries are available on all target platforms. The correct way of using XAudio in windows is to abstract the audio engine in OpenTomb so that on windows it uses XAudio directly, while using OpenAL on all other platforms. Patching XAudio into OpenAL is surely the wrong way, as they both are abstraction layers with the same underlaying layer. And well, if developers cannot even install a needed lib, then they are not worth the title "developer".

Sorry, but I still haven't seen any reasonable argument that let me even *consider* to leave the code in.

**Comment by vvs-** [comment]
> And well, if developers cannot even install a needed lib, then they are not worth the title "developer".

That's a very strong statement that I'd like to avoid. We don't need to alienate people without system administration experience. I know many *professional* Java developers working on Windows who don't even know what CMD.EXE is. Don't lough - it's true.

**Comment by stohrendorf** [comment]
I have my very strong opinion about Java "developers", and I'm not laughing, as I know that's true.

But we're talking about C/C++ developers here. If you want to attract people that do not know how to use/build/install external libs, this project is going to be doomed, as such people usually write code that's ... well, weird and not very error-prone.

Please don't think I'm only ranting. But the need for including the libs in-source comes from reasons that will lead to great problems in the long term.

**Comment by cochrane** [comment]
I don't mean to be rude here, but the first and only time we've had absolutely massive problems with libraries was when someone decided to break all existing processes for dealing with them because it looked cleaner. I like insulting people as much as anybody, but in the end it has rarely, if ever, solved problems.

If you think you have a solution that works better than what we had, great, show it to me. So far I haven't seen it.

**Comment by vvs-** [comment]
To be fair, nobody openly opposed that move. So, I think it's a scapegoating to blame the author of the pull request.

**Comment by stohrendorf** [comment]
It's pretty simple, but requires some work: Implement *two* audio engines, one for OpenAL, one for XAudio. This needs an abstraction/indirection layer in OpenTomb, but it's the only clean way, because patching XAudio support into OpenAL is like patching DirectX support into OpenGL.

**Comment by cochrane** [comment]
> To be fair, nobody openly opposed that move. So, I think it's a scapegoating to blame the author of the pull request.

You're right. @stohrendorf : Sorry.

> It's pretty simple, but requires some work: Implement two audio engines, one for OpenAL, one for XAudio. This needs an abstraction/indirection layer in OpenTomb, but it's the only clean way, because patching XAudio support into OpenAL is like patching DirectX support into OpenGL.

I don't really understand your objection here. Yeah, it is a bit weird, but it is also code that is already tested and works well. You propose to replace that well-tested code and replace it with a huge new thing that needs to be written and tested again and which will expose an interface that nobody is familiar with, for no actual benefit other than that a description of it sounds nicer.

I understand the desire for clean solutions, but maintenance cost is a real thing, too. Now, if you want to write that translation layer, be my guest. But I just don't see the benefit.

**Comment by stohrendorf** [comment]
Then there's FMOD Ex; it isn't open source, but a fairly advanced sound engine for win/mac/linux. A colleague who worked for a game developer told me about that.

**Comment by vvs-** [comment]
I'd second the maintenance concerns.

> Then there's FMOD

Wouldn't that be incompatible with OpenTomb license?

**Comment by vvs-** [comment]
To add to my previous post. There are such big projects as OpenStack and Xen which use their own development environment (DevStack and Raisin respectively). They have just the same concerns about consistency of their development environment for the developers and easy maintenance. Though, it has a big cost of designing and constructing such reference environment in the first place.

But having all reference libraries as an optional git submodules used by the developers can ease the pain considerably.

**Comment by vvs-** [comment]
I think it requires some elaboration.

The idea is to move all external libraries into submodules to make it optional. Those who require special library versions should *explicitly* checkout the submodules. That means cross-compiling or some platform specific versions which should be hosted in the opentomb organization. Others shouldn't do anything special and use system libraries by default.

We can even have different branches for that, i.e. development or stable which could differentiate between debug and release specific libraries.

**Comment by ablepharus** [comment]
Why not use the DirectSound backend of openal-soft on Windows and leave XAudio alone for now? If you want to have XAudio, work on openal-soft.

**Comment by Lwmte** [comment]
In fact, I got reply from Chris Robinson (who works on OpenAL Soft), and he told me that there's no benefit in XAudio2, when compared to mmdevapi (actually, people say that XAudio2 is an abstraction layer over mmdevapi), and OpenAL Soft already uses mmdevapi as a backend for Windows systems since Vista. So it's no more actual.

**Comment by stohrendorf** [comment]
And the mac-issues with missing extensions could also be solved with b3cb74737a542eef011b6ceea8ef62e059b7a267, as I had missing extensions here on windows, too, until I implemented probing all output devices for EFX support.

**Comment by Lwmte** [comment]
By the way, can we store OpenAL Soft as a reference too? OpenTomb makes no use of native SDL audio features, and now, when XAudio2/mmdevapi business is sorted out, we can use official version without modifications.
P.S.: Chris also said to me that using SDL as a backend with OpenAL is unsafe, even in SDL2. There is a risk that OpenAL may still be shutting down while SDL is already destroyed on exit, which is dirty.

If you need a link to OpenAL git repo, here is it: http://repo.or.cz/w/openal-soft.git

**Comment by Lwmte** [comment]
Also, what about SDL2, SDL_Image, zlib and Lua? I don't remember whole story, but why these were left as links to pre-compiled libs, if we decided to convert everything to references? Or there is a reason for such setup?

**Comment by vvs-** [comment]
As far as I remember, you and TeslaRus said that the only patched libraries were bullet and openal. Why freetype is there I don't remember. It was decided that it's better to use system libraries for everything else. Actually, I think that developers and users should have different setups.

**Comment by Lwmte** [comment]
Well, currently we're not using patched Bullet anymore (by the way, Lua was also patched, but it is now referenced as external library, not as submodule), and seems we don't need patched OpenAL either after all this discussion about XAudio. So effectively, only patched library we currently need is stohrendorf's LuaState fork (ironically, as Steffen proposed removing patched libs in the first place).

Cochrane had problems with stock OpenAL for Mac (I have no idea what's the version, probably it's a fork of last open-source Creative OpenAL, which should be ridiculously ancient), what he really needs is **any** version of OpenAL **Soft**, and there is no real difference between submodule reference to current git repo of OpenAL Soft and bundled patched version. Only difference is backend.

So... I believe it's time to finally decide how we manage libs. If we've decided to use submodules, we should get rid of system libs completely (meaning - SDL2, SDL_Image, zlib and Lua should be also added as submodules... ah, I also forgot about libsndfile and ogg/vorbis). Right?

**Comment by Lwmte** [comment]
And I'll say again - I believe there is an ongoing issue with different includes of Bullet modules, as I have seen ` <> ` and ` "" ` links to them roaming around.

**Comment by vvs-** [comment]
I don't see why it should be either one or the other. Why not both? You can keep only patched libs as submodules and rely on the system libs for everything else. I should point out that if OpenTomb will ever get packaged for distribution, then it *should* rely on system libraries.

As for includes I think it's better keep them global as `<>`.

**Comment by Lwmte** [comment]
Ok then, basically, we need **only** LuaState as reference then.

**Comment by vvs-** [comment]
Sure. We can always add anything if needed.

**Comment by vvs-** [comment]
BTW, LuaState is header only. It doesn't need runtime.

**Comment by Lwmte** [comment]
But then we return to the problem described by Cochrane in the first post - what if someone has different library version? I must say that so far it wasn't an issue for me, cause I never encountered compile problems because of wrong lib versions. Only ogg/vorbis were picky about particular flag settings (` fseeko64 ` function), but these were integrated all along (I should confess, it was me integrating ogg/vorbis sources, because I couldn't find pre-compiled lib binaries for MinGW).

**Comment by Lwmte** [comment]
And, as I said earlier, I'm also in favor of this approach (system libs). Only headache for me would be compiling ogg/vorbis.

**Comment by vvs-** [comment]
> what if someone has different library version?

We need to choose something. And every choice will have its pros and cons. We can change our decision later. Right now it's not a primary concern.

**Comment by Lwmte** [comment]
For me it's the primary concern now, as I want to finally build this source code! :smile:

**Comment by vvs-** [comment]
If I understood correctly, you'll need to build all the libraries anyway no matter if they in-tree or outside.

**Comment by Lwmte** [comment]
Yes, and building libraries themselves is something I never did before... Actually I just googled for compiled .a/.la files and just put them into /lib directory, and headers into /include directory! But this won't work for ogg/vorbis and bullet, because there's no pre-compiled library binaries on the net (except outdated versions).

**Comment by vvs-** [comment]
If there is CMakeFile somewhere then consider it done. If not then CygWin could help. In any case there should be a tutorials somewhere on the Net for building it under Windows. If nothing will work just ask here and may be someone will have an answer.

**Comment by Lwmte** [comment]
I have successfully built both Bullet and Ogg/Vorbis. It was actually piece of cake, no errors or warnings, only thing I needed to do is to change build type in CodeBlocks from "commands only" to "static library". Now I'll try to build latest OpenAL Soft! :sweat_smile:

**Comment by Lwmte** [comment]
Ok, I have built all the libraries, and now I'm trying to compile engine using CMake-generated CodeBlocks project. I should also note that I've removed integrated OpenAL and put freshly compiled libs and includes into my minGW folder. The problem is, CMake links to OpenAL **dll** from Windows system directory, but I want it to link to my static library from minGW folder. How can I do this?

**Comment by Lwmte** [comment]
It's a miracle! I have successfully compiled it without warnings and errors! How it's possible? Every time people introduced something new before, I had to break my head trying to find what's missing, like flags, includes, build options, etc. But now it looks like some kind of CMake magic... I think I'll use CMake only for now.

However... Seems that OpenAL Soft isn't the only one which tries to link dynamically... Glew also wants to be a dll, how can I prevent that? Is there any command for CMake which forces certain lib to be identified and linked as static?

**Comment by cochrane** [comment]
Sorry for taking soong. Yes, my problem with OpenAL is that the version included with Mac OS X is ancient. Any version of OpenAL Soft should work just fine for me.

**Comment by cochrane** [comment]
*so long (can't edit my comments on the mobile version of Github)

**Comment by Lwmte** [comment]
Yes, then I think we should just leave OpenAL Soft and LuaState as submodules (OpenAL Soft for Mac only).
However, how can we use OpenAL as static lib in win/linux, while using it as submodule on mac? How to manage includes in such case?

**Comment by Lwmte** [comment]
P.S.: Just found CMake command BUILD_SHARED_LIBS, will it prevent linking dynamically?

**Comment by Lwmte** [comment]
...However, on second thought, I had to put these DLLs in the same directory as binary:

* glew32.dll
* libjpeg-9.dll
* libpng16-16.dll
* libsndfile-1.dll
* libtiff-5.dll
* libwebp-4.dll
* lua53.dll
* SDL2_image.dll (magically, SDL2.dll isn't required)
* zlib1.dll

Don't you think it's too much?

**Comment by vvs-** [comment]
Oh, libpng, libtiff and libjpeg are dependencies too. It's a standard set of libraries which seems to always be on any Linux system, so I didn't even notice that they are required. That shows that to build a complete set of requirements we need to build it on all supported systems. And I think cross-compiling should be include too.

**Comment by vvs-** [comment]
> Just found CMake command BUILD_SHARED_LIBS, will it prevent linking dynamically?

"*   BUILD_SHARED_LIBS
       Global flag to cause add_library to create shared libraries if on.

       If present and true, this will cause all libraries to be built shared unless the library was explicitly
       added as a static library.  This variable is often added to projects as an OPTION so that each user  of
       a project can decide if they want to build the project using shared or static libraries.*"

**Comment by Lwmte** [comment]
This is exactly what we need.
Although, I tried to modify CMakeLists.txt to make static libraries, but it had no effect... I think someone with better knowledge of CMake should implement this.

**Comment by vvs-** [comment]
But we already identified this bugs (at least partially):

>Lara looks like abomination,

That's #143. Not only Lara, just any animation.

> textures are messed up,

That's #124.

> entities are missing, etc.)

That's #145.

**Comment by vvs-** [comment]
> This is exactly what we need.

No it isn't. It controls building of shared libraries but not static.

**Comment by vvs-** [comment]
> Please check recent changes - I have removed Bullet and freetype

Yes, now CMakeLists is broken. The commands to find them were in 3rdparty/CMakeList.

**Comment by Lwmte** [comment]
Sorry for this, I thought this CMakeLists file was only used if bullet/freetype were added as submodules.
Also I wasn't aware of current engine state, as it's first time I've compiled it since the big merge. I just saw your screenshot and thought that texture and animation issues were already fixed.
Regarding CMake static/dynamic library linking - then how can we explicitly tell it that all live must be static?

**Comment by Lwmte** [comment]
Sorry, libs, not live. Stupid autocorrection.

**Comment by vvs-** [comment]
> I thought this CMakeLists file was only used if bullet/freetype were added as submodules.

And I think that's exactly why it was a wrong decision to put it there.

>  I just saw your screenshot and thought that texture and animation issues were already fixed.

We need them fixed and Steffen has no knowledge of how the old version actually worked. So, we need help from all the developers with that knowledge.

> Regarding CMake static/dynamic library linking - then how can we explicitly tell it that all live must be static?

I don't know yet. We'll need to consult the manual :smile:

**Comment by Lwmte** [comment]
Ok then, I'll see what I can do when I get home :smiley:

**Comment by vvs-** [comment]
`add_library` command allows to specify which type of library should be built.

"*If no type is given explicitly the type is STATIC or SHARED based on whether the current value of the  variable  BUILD_SHARED_LIBS  is ON.*"

So, it builds static libraries by default. And the command `find_library` can be used to add existing libraries to the linker list.

**Comment by vvs-** [comment]
Here's an example: 
```
add_library(l SHARED l.c)
add_executable(e e.c)
target_link_libraries(e l)
```
This project includes one executable and one library. It compiles both, builds shared library and links executable dynamically. To build static library and link statically, just remove `SHARED`.

**Comment by vvs-** [comment]
And I see that there is a convenient function `trylinklib` in OpenTomb already. In the current CMakeLists it tries to find shared versions several libs but that could be changed. But such change should be encapsulated in a platform specific manner because not every platform has static versions of all libraries.

**Comment by stohrendorf** [comment]
Proposal (and the only sane way): Only list libs that we require *directly*, i.e. *at first hand*. Transient dependencies are handled by the respective package maintainers, and if one wants to compile these libs by hand... well, these libs have listed their deps for themselves.

**Comment by vvs-** [comment]
That's a good idea for unices. But I doubt it would please developers on Windows.

**Comment by stohrendorf** [comment]
Developers on Windows are *always* unpleased, unless they use VS, but then they are only pleased if everybody develops *only* on windows using VS. By the way, OpenTomb already uses C++11 features that VS still doesn't support. And MSYS2 has a really neat package manager (`pacman`) that does its job really good, and let me set up a dev env on windows unbelievable easy compared to trying to setting it up using VS.

**Comment by vvs-** [comment]
That means we are back to [discussion](https://github.com/opentomb/OpenTomb/issues/115#issuecomment-118396297) about good development environment for OpenTomb. Then Windows developers should be pleased once and for all :wink:

**Comment by Lwmte** [comment]
Oh, there is MSYS2 already? :open_mouth: 

Talking about dependencies... Are there wrong entries currently in CMakeLists.txt? I'm worrying primarily about sound libs. sndfile requires ogg/vorbis libs explicitly, or they count as dependencies? If so, we can safely remove everything except linking sndfile itself.

**Comment by vvs-** [comment]
As far as I understand, cmake counts only explicit library dependencies listed in CMakeLists. At least I was not able to link my test example above without it.

**Comment by stohrendorf** [comment]
sndfile requires ogg/vorbis, but it needs the target program to link against them. That's also stated on the homepage.

**Comment by cochrane** [comment]
Sorry I missed out on the discussion over the last couple of days.

> Proposal (and the only sane way): Only list libs that we require directly, i.e. at first hand. Transient dependencies are handled by the respective package maintainers, and if one wants to compile these libs by hand... well, these libs have listed their deps for themselves.

Who are the package maintainers in your scenario? As far as I understand it, I'm the one for the Mac version, almost everybody else is managing the Windows version. There is no external package maintainer role that we can shift the responsibility to.

This proposal, and all the related things, mean that I either have to set up a zip file with all the libraries required for Mac somewhere and tell every Mac developer to download it in addition to the code. Alternatively, I can just say "You have to jump through at least <N> hoops if you want to help us", meaning telling them to install random versions of the library themselves. That may indeed be the only practical way on Linux, where everyone has a randomly different operating system, but I have never heard of it as best practice anywhere else.

**Comment by stohrendorf** [comment]
Does that mean that Mac OS X handles packages like windows, i.e. every app has its own copy of the libs? I thought there was a dependency system like in DEB or RPM, too.

**Comment by cochrane** [comment]
No, there is no dependency system like that. It's like Windows in that regard.

Unlike Windows, it's easy to install your own dependency manager, and there are several around, but it's still only something for power users. We can't expect normal people to install OpenTomb that way. Or at least I don't want that, I want it to be a normal OS X app, because that's a much better user experience.

**Comment by stohrendorf** [comment]
So what do you propose then? I can only speak for windows and linux.

**Comment by cochrane** [comment]
What I said at the start: Include the source code as external references, or the pre-built libraries (far less preferred).

I realize this isn't perfect for Linux, so I'm open for better suggestions. I think the external references would work well on Windows, too.

**Comment by stohrendorf** [comment]
What about putting a `DEPENDENCIES` file into the source tree, with links to archives with the correct version? I'd personally prefer that instead of using submodules, because then you are less tempted to patch around in the libs.

**Comment by cochrane** [comment]
Hm… I don't like it at all, because it's annoying, but it would certainly work. With the list in the Wiki, we do have most of the work for that done already. Still, it's annoying.

I think I'm going to do some research on how other open-source projects with a Mac version (as opposed to Mac-only open source projects) handle this issue, and report back.

**Comment by stohrendorf** [comment]
Are there any points unresolved? I'd like to close this, as the system as it is right now seems to work.

**Comment by cochrane** [comment]
I still haven't had time to do the researcher ruined above, so I'd prefer to keep this open

**Comment by cochrane** [comment]
Damn autocorrect. You know what I mean.

**Comment by stohrendorf** [comment]
I have seen a bundle script for app bundles in cmake, which (according to what I understood) should do all the bundling automatically: http://www.cmake.org/cmake/help/v2.8.12/cmake.html#module:BundleUtilities

**Comment by Lwmte** [comment]
I still don't get an idea of referencing submodule of Bullet in CMake, while we don't have one in actual source code.

**Comment by vvs-** [comment]
You might be missed [this](https://github.com/opentomb/OpenTomb/issues/210#issuecomment-124846829) then.

**Comment by stohrendorf** [comment]
Probably the best way to handle bullet is this:
* Check out bullet 2.83.5 or later into $DIR.
* Then `mkdir -p $DIR/build && cd $DIR/build && cmake -DCMAKE_INSTALL_PREFIX=$HOME/some/dir .. && make && make install` to install it into `$HOME/some/dir`
* Now get OpenTomb to build using that local installation using `cmake -DBULLET_ROOT=$HOME/some/dir ...`

**Comment by vvs-** [comment]
This is too heavy considering that nobody will ever use it except OpenTomb. And it builds too much stuff that OpenTomb doesn't use.

**Comment by cochrane** [comment]
No offense, but the above is exactly the reason why I think the current situation is not an acceptable resolution to this issue. 

So I did the research, and among serious projects I found two main ways: One is to just out the requirements in a readme, and let developers deal with it, used e.g. By Wings3D. The other is to include the source code either directly (Blender) or in a separate repository (Ogre3D). Some put everything into their SVN repository, which I think is just ugly. 

The current "just put it in a readme" approach is okay on systems with a package manager, where installing the prerequisites is fast and easy. In reality, the systems used by most people and most developers here are ones that don't have that. Installing these requirements by hand is annoying as hell, especially when people keep adding new ones.

So I think we absolutely need the libraries in one place in our project. Can be as submodules so you don't patch around in them (have you ever tried patching around in a submodule? You'll realize the error of your ways very quickly).

**Comment by stohrendorf** [comment]
@vvs- Then the cmake command line for bullet needs to be extended with `-DBUILD_BULLET3=OFF -DBUILD_CPU_DEMOS=OFF -DBUILD_EXTRAS=OFF -DBUILD_OPENGL3_DEMOS=OFF -DBUILD_SHARED_LIBS=OFF -DBUILD_UNIT_TESTS=OFF`. That's what works in windows, and it's a minimal build. I'm just concerned that the other approach of removing the system provided libs leads to a system that's too specialized.

@cochrane I could live with bullet as a submodule, as long as it's a bit abstracted. I think of something like a "pre-configure step" to separately build bullet in-tree, with the commands state above. I.e., when doing the first cmake on OpenTomb, bullet would be automatically built, and then used. I'll try that and maybe put on a merge request later.

**Comment by cochrane** [comment]
Why only bullet?

**Comment by stohrendorf** [comment]
The pull request is open (#247), and is a "works for me version" (TM). Should be easily extensible to other libs that use cmake.

**Comment by vvs-** [comment]
> removing the system provided libs leads to a system that's too specialized.

And if you don't then you are risking to inadvertently bring different versions of headers or even libraries.

**Comment by stohrendorf** [comment]
If the compilers don't respect local includes over system includes, then the compilers are not standard-compliant (and crap). The approach I tried in #247 should prefer local includes and libs from the locally built bullet over the ones provided by the system.

**Comment by vvs-** [comment]
That works only if the project file doesn't have bugs. And if it does that leads to a very obscure brokenness which is hard to track down. I dealt with such cases numerous times and can assure that they are quite common.

**Comment by stohrendorf** [comment]
So you're saying that #247 probably doesn't solve any problems, but instead makes them more obscure?

**Comment by vvs-** [comment]
I never said anything about #247. I even didn't look at it yet. What I meant is that it's very unreliable to count on the paths and that they will be always in correct order. Better take no chances and never install several versions of the same library unless you have some sort of a chroot or a container.

**Comment by stohrendorf** [comment]
Sorry, I misunderstood you, and I agree with you. But from looking at the code, we don't use the `bullet/` path prefix for the includes anymore, which means that the include root path must be specified on the command line, so at least here is no way the compiler may confuse the pathes. And the bullet libs are qualified using their absolute file pathes, so here should be no problem, too. (I'm primarily promoting #247 here, but it's valid for the current `master` branch as well.)

OpenAL is handled the same way; e.g., my CMakeCache.txt contains:
```
OPENAL_INCLUDE_DIR:PATH=/usr/include/AL
OPENAL_LIBRARY:FILEPATH=/usr/lib64/libopenal.so

BULLET_COLLISION_LIBRARY:FILEPATH=/home/sto/devel/OpenTomb-bullet/3rdparty/bullet3/install/lib/libBulletCollision.a
BULLET_COLLISION_LIBRARY_DEBUG:FILEPATH=BULLET_COLLISION_LIBRARY_DEBUG-NOTFOUND
BULLET_DYNAMICS_LIBRARY:FILEPATH=/home/sto/devel/OpenTomb-bullet/3rdparty/bullet3/install/lib/libBulletDynamics.a
BULLET_DYNAMICS_LIBRARY_DEBUG:FILEPATH=BULLET_DYNAMICS_LIBRARY_DEBUG-NOTFOUND
BULLET_INCLUDE_DIR:PATH=/home/sto/devel/OpenTomb-bullet/3rdparty/bullet3/install/include/bullet
BULLET_MATH_LIBRARY:FILEPATH=/home/sto/devel/OpenTomb-bullet/3rdparty/bullet3/install/lib/libLinearMath.a
BULLET_MATH_LIBRARY_DEBUG:FILEPATH=BULLET_MATH_LIBRARY_DEBUG-NOTFOUND
BULLET_SOFTBODY_LIBRARY:FILEPATH=/home/sto/devel/OpenTomb-bullet/3rdparty/bullet3/install/lib/libBulletSoftBody.a
BULLET_SOFTBODY_LIBRARY_DEBUG:FILEPATH=BULLET_SOFTBODY_LIBRARY_DEBUG-NOTFOUND
```

(By the way, I *have* the debug versions of Bullet installed in the system, but CMake hasn't used them, since BULLET_ROOT was used.)

**Comment by vvs-** [comment]
Of course, I don't say that it will necessary happen, but the risk is still there. In fact, the old CMake project included different Bullet paths *twice* and nobody even noticed that. So, it's 50/50 and I don't like that risk. I still suggest to always remove any alternate versions before compiling in-tree.

**Comment by stohrendorf** [comment]
I think this discussion became somewhat subjective...

I agree that multiple installations can be a problem, and I think your aproach of removing the system libs is a *good* suggestion if something seems wrong, but I don't believe that the chance of such a misconfigured build environment is at 50%. (Hell, I *believe* that #247 works in over 90% of the cases, and I *hope* I'm not wrong o.O)

**Comment by Sur3** [comment]
I don't see a reason to include code from libraries in an own project, cmake should handle finding the system libraries if installed and if not warn about it missing. If modifications to external libraries are really needed you could clone the external source in an own git and set a tag to that and also commit a pull request upstream. And also an option to use the system library instead in cmake should be included in that case, probably hinting what features are disabled if that is used instead of the bundled version. But changing libraries without upstream pull requests or possibility to use system libraries instead is always a bad idea; I started developing supertuxkart recently and they have their own version of a Irrlicht engine which was pretty confusing because the changes are not documented well and if searching for Irrlicht tutorial they don't fully apply to that code.. The same problem goes for every project duplicating libraries making local changes.

**Comment by vvs-** [comment]
I agree with your conclusion. But an amateur projects just don't have a required discipline or incentive to follow recommended best practices. That's just a fact with which we have to live on.
